### DISCUSSION TITLE
Add a timeout to chunked upload requests, so it tries again.

### DISCUSSION THREAD

**Discussion by Piero512** [rejetto/hfs#844] [General]
Hello

I was looking into this fileserver and it's very featured. I like the out of box experience.

My suggestion is to attach a timeout for the put requests used to upload files, so we can try it again later, instead of getting the upload just hang.

![image describing an upload that hangs, with devtools showing the issued requests.](https://github.com/user-attachments/assets/1dca28cf-9d1b-452f-bc52-e04ff0d4cb1e)

Notice in devtools that thre is a hanging request, which I think CF just made it, but the client code doesn't seem to try cancelling the request if it takes too long and just reissuing it.

**Comment by rejetto** [comment]
hi Piero!
how many times did you see this problem?
is there a way for me to reproduce this problem?
how long did you wait to determine the request is stuck?
and what's exactly the rule you suggest to implement? is it upload speed zero? for how long?

**Reply by Piero512** [reply]
I can reproduce it consistently, by trying to upload that 4GB file when running the server on Windows 10/Windows Server 2019 standard edition, through CF reverse proxy. from another computer from Europe (I use those database backups to diagnose problems and decided that I did not wanted to upload those DBs to an external service).

I literally have tried this whole day to get it to work. so the wait was more than 5 minutes, and you can see the chunks on the screenshot took way less. Tried experimenting with different chunk sizes since it was recommended to set it to 100mb for CF free.

Probably the easiest way to try it would be to try and run an instance in any supported device, run cloudflared to run the service, and try uploading a huge file.

I wish it would be less of a fight to just request a port forward in the place I run the server (my job). 

TBH, I don't know what would be a good rule. Have you ever heard of The Upload Server [protocol](https://tus.io/)? I hope they have solved the issue already. It has different backends, so maybe it can be plugged in on the VFS that HFS builds for the client?

**Comment by rejetto** [comment]
i didn't know about this TUS project. Sounds interesting, but nothing i can consider in the immediate moment.
from your screeshot, i understand some chunks are correctly uploaded, but at some point it stops.
i guess you don't know if it stops in the middle of a chunk, or at the very start of it.
i know there are other CF users who never reported any problem with uploads this big. I wonder what's different with your case.

1. Your suggestion to cancel and reattempt makes sense, but we don't know if that would actually work, right?
2. What happens if you stop the upload (even with a brutal reload of the page), and attempt another upload? It should offer you to resume. Does it progress this way? Does it stop after some more chunks?

**Reply by Piero512** [reply]
The resume functionality is not stable enough, ie, it fails when the chunk size is low, as it assumes that you didn't want to resume and it's also manual. Maybe the resume functionality should check before issuing the upload request if you want to resume, instead of doing it with the pop-up while uploading

**Comment by rejetto** [comment]
the reason why it uploads while waiting for the user to answer the dialog, is to allow long unattended uploads without stopping.

Yes, i see that it is not handled correctly if the second chunk starts before the user has answered the resume-confirmation-dialog. Thanks for pointing that out. I will work on that.

**Comment by rejetto** [comment]
i both fixed the resume bug (hopefully), and implemented a retry mechanism in case the upload is stuck.
i made my tests by simulating the hanging (twice in a row). I can't test with CF at the moment.
would you be available to make a test?
i can give you a previous version
https://github.com/user-attachments/files/18291455/hfs-windows-x64-0.55.2-beta1.zip
you can easily install it by right-clicking on "check for updates" in admin-panel/home.

**Reply by Piero512** [reply]
Hmm, I would be able to try the exact setup when the new year holiday ends in my country, on January 6th.

Probably today in the afternoon I could try the build in my personal computers.

**Comment by rejetto** [comment]
if you didn't yet, use this
https://github.com/user-attachments/files/18293786/hfs-windows-x64-0.55.2-beta2.zip
as it includes more fixes

**Comment by rejetto** [comment]
you don't need to right-click anymore.
Just normal click, after enabling the "include beta versions" option.

**Comment by Piero512** [comment]
Hey, so I have been trying HFS 0.55.2 and it seems to have improved it? I sometimes see cancelled requests but maybe our internet is that slow.

**Comment by rejetto** [comment]
i had a closer look to TUS and it seems (from the documentation) to violate a principle i had in mind designing my upload system: to NOT slow down with many small files. Of course "small" depends on the speed of the connection, that normally you cannot determine ahead, so any file is possibly small (in principle, at least).
HFS always starts uploading without waiting to know if it's possible to resume. If the file is small, this is much quicker: asking could take the same time as it takes to upload. If you have a folder with many small files, asking for each of them can greatly slow down the whole process, up to doubling the time, i guess.
And i'm not even considered a factor that is greatly worsening the problem: since i cannot assume the partial file is good for resuming, as the origin file may have changed in the meanwhile, i have to ask the user, which is even much slower at deciding. 
I may add a hashing system (on-demand), a burden for the server but may be worth it.

This is not the case of all systems: if you are dealing with media (audio/video/pictures) you can assume your files are never small. Sadly this doesn't apply to HFS.
This doesn't necessarily mean I made the best decision. Perhaps I improved 1% of the cases while inadvertently worsening the situation for the remaining 99%. I cannot judge by single experiences, and I don't have statistics.
